File: app/src/main/java/com/fruits/fruits/FruitsAdapter.java
package com.fruits.fruits;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.fruits.R;
import com.fruits.fruits.FruitsListItemClickListener;
import com.fruits.model.Fruit;

import java.util.ArrayList;
import java.util.List;

public class FruitsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder>   {
    List<Fruit> fruits = new ArrayList<>();

    private Context context;
    private FruitsListItemClickListener fruitClickListener;

    public FruitsAdapter(Context context, FruitsListItemClickListener fruitClickListener){
        this.context = context;
        this.fruitClickListener = fruitClickListener;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(context);
        View view = layoutInflater.inflate(R.layout.list_item_fruit, parent, false);

        FruitViewHolder fruitViewHolder = new FruitViewHolder(view);

        return fruitViewHolder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        final Fruit fruit = fruits.get(position);
        ((FruitViewHolder)holder).bind(fruit);

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                fruitClickListener.itemClicked(fruit);
            }
        });
    }

    @Override
    public int getItemCount() { return fruits.size();}

    public void update(List<Fruit> fruits) {
        this.fruits.clear();
        this.fruits.addAll(fruits);
        notifyDataSetChanged();
    }

    static class FruitViewHolder extends RecyclerView.ViewHolder {

        public FruitViewHolder(View itemView) {
            super(itemView);
        }

        public void bind(Fruit fruit){
            TextView fruit_type = itemView.findViewById(R.id.fruit_type);
            fruit_type.setText(fruit.getType());
        }
    }
}

File: app/src/main/java/com/fruits/fruits/FruitsAdapter.kt
package com.fruits.fruits;

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.fruits.R
import com.fruits.model.Fruit
import java.util.*

class FruitsAdapter(context: Context, fruitClickListener: FruitsListItemClickListener) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    var fruits: MutableList<Fruit> = ArrayList()

    private val context: Context = context;
    private val fruitClickListener: FruitsListItemClickListener = fruitClickListener;

    @Override
    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): RecyclerView.ViewHolder {

        val layoutInflater: LayoutInflater = LayoutInflater.from(context);
        val view: View = layoutInflater.inflate(R.layout.list_item_fruit, parent, false);

        val fruitViewHolder = FruitViewHolder(view);

        return fruitViewHolder;
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder?, position: Int) {
        val fruit: Fruit = fruits.get(position);
        (holder as FruitViewHolder).bind(fruit);

        holder.itemView.setOnClickListener(View.OnClickListener { view ->
            fruitClickListener.itemClicked(fruit);
        });
    }

    @Override
    override fun getItemCount(): Int = fruits.size

    fun update(fruits: MutableList<Fruit>) {
        this.fruits.clear();
        this.fruits.addAll(fruits);
        notifyDataSetChanged();
    }

    companion object {
        class FruitViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

            fun bind(fruit: Fruit) {
                val fruit_type: TextView = itemView.findViewById(R.id.fruit_type);
                fruit_type.setText(fruit.type);
            }
        }
    }
}
